pkg/apis/migration/migration.go

```diff
@@ -92,46 +92,49 @@ const (
 	CRDPlural = "cnsvspherevolumemigrations"
 )
 
-// onceForVolumeMigrationService is used for initializing the VolumeMigrationService singleton.
-var onceForVolumeMigrationService sync.Once
-
-// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
-var volumeMigrationInstance *volumeMigration
-
-var volumeMigrationServiceInitErr error
+var (
+	// volumeMigrationInstance is instance of volumeMigration and implements interface for VolumeMigrationService
+	volumeMigrationInstance *volumeMigration
+	// volumeMigrationInstanceLock is used for handling race conditions during read, write on volumeMigrationInstance
+	volumeMigrationInstanceLock sync.Mutex
+)
 
 // GetVolumeMigrationService returns the singleton VolumeMigrationService
 func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config) (VolumeMigrationService, error) {
 	log := logger.GetLogger(ctx)
-	onceForVolumeMigrationService.Do(func() {
+	volumeMigrationInstanceLock.Lock()
+	defer volumeMigrationInstanceLock.Unlock()
+	if volumeMigrationInstance != nil {
 		log.Info("Initializing volume migration service...")
 		volumeMigrationInstance = &volumeMigration{
 			volumePathToVolumeID: sync.Map{},
 			volumeManager:        volumeManager,
 			cnsConfig:            cnsConfig,
 		}
-		volumeMigrationServiceInitErr = k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
+		volumeMigrationServiceInitErr := k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
 			reflect.TypeOf(migrationv1alpha1.CnsVSphereVolumeMigration{}).Name(), migrationv1alpha1.SchemeGroupVersion.Group, migrationv1alpha1.SchemeGroupVersion.Version, apiextensionsv1beta1.ClusterScoped)
 		if volumeMigrationServiceInitErr != nil {
+			volumeMigrationInstance = nil
```

Comment:
Instead of this, how about we have a boolean variable(say isInitialized) set to true only in the success case? Line#107 can check for this flag to decide whether volumeMigrationInstance is initialized or not.

Reply:
Do you think this is still needed after adding double checked locking ?

Reply:
Yes. I would have just defined a boolean that tells if volumeMigrationInstance is initialized or not. But this is subjective as we can also set volumeMigrationInstance to nil and it should work.

Reply:
Keeping this as it is for now